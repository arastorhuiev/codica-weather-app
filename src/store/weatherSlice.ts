import axios from 'axios';

import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { Status, WeatherState } from './types';

const baseFetchDataByCityName = async (city: string) => {
  const response = await axios.get(
    `https://api.openweathermap.org/data/2.5/weather?q=${city}&units=metric&appid=${process.env.REACT_APP_OPEN_WEATHER_API_KEY}`,
  );

  return response.data;
};

export const fetchWeatherByCityName = createAsyncThunk(
  'weather/fetchWeatherByLatLng',
  baseFetchDataByCityName,
);

export const reloadDataByCityName = createAsyncThunk(
  'weather/reloadDataByCityName',
  baseFetchDataByCityName,
);

const initialState: WeatherState = {
  status: Status.LOADING,
  statusReloadedCity: null,
  data: [],
};

const weatherSlice = createSlice({
  name: 'weather',
  initialState,
  reducers: {
    removeDataByCityId(state, action) {
      const cityId = action.payload;
      const filtered = state.data.filter((item) => item.id !== cityId);
      state.data = filtered;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchWeatherByCityName.pending, (state) => {
      state.status = Status.LOADING;
    });
    builder.addCase(fetchWeatherByCityName.fulfilled, (state, action) => {
      state.data.push(action.payload);
      state.status = Status.SUCCESS;
    });
    builder.addCase(fetchWeatherByCityName.rejected, (state) => {
      state.status = Status.ERROR;
    });
    builder.addCase(reloadDataByCityName.pending, (state) => {
      state.status = Status.LOADING;
    });
    builder.addCase(reloadDataByCityName.fulfilled, (state, action) => {
      const cityName = action.payload.name;
      const cityData = action.payload;
      const cityIndex = state.data.findIndex((item) => item.name === cityName);
      state.data.splice(cityIndex, 1, cityData);
      state.status = Status.SUCCESS;
      state.statusReloadedCity = Status.SUCCESS;
    });
    builder.addCase(reloadDataByCityName.rejected, (state) => {
      state.status = Status.ERROR;
    });
  },
});

export const { removeDataByCityId } = weatherSlice.actions;
export const weatherReducer = weatherSlice.reducer;
