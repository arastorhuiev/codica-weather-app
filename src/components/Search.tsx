import { Box, TextField } from '@mui/material';
import { useState } from 'react';
import { useAppDispatch } from '../store/hooks';
import { fetchWeatherByCityName } from '../store/weatherSlice';

export function Search() {
  const [currentCity, setCurrentCity] = useState<string>('');

  const dispatch = useAppDispatch();

  const handleChangeByValue = (event: React.ChangeEvent<HTMLInputElement>) => {
    setCurrentCity(event.target.value);
  };

  const dispatchByEnterKey = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') {
      dispatch(fetchWeatherByCityName(currentCity));
    }
  };

  return (
    <Box>
      <TextField
        label='Enter city name'
        value={currentCity}
        onChange={handleChangeByValue}
        onKeyPress={dispatchByEnterKey}
        variant='standard'
        fullWidth
      />
    </Box>
  );
}
