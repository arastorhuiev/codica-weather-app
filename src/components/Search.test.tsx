import { render, screen, fireEvent } from '@testing-library/react';
import * as reduxHooks from 'react-redux';
import * as actions from '../store/weather/weatherSlice';
import { Search } from './Search';

jest.mock('react-redux');

const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch');

describe('Search', () => {
  it('should create Search', () => {
    mockedDispatch.mockReturnValue(jest.fn());
    const view = render(<Search />);

    expect(view).toMatchSnapshot();
  });

  // it('should dispatch action', () => {
  //   const dispatch = jest.fn();
  //   mockedDispatch.mockReturnValue(dispatch);

  //   const mockedFetchByEnterKey = jest.spyOn(actions, 'fetchWeatherByCityName');

  //   render(<Search />);

  //   fireEvent.keyPress(screen.getByRole('textfield'));

  //   expect(dispatch).toHaveBeenCalledTimes(1);
  //   expect(mockedFetchByEnterKey).toHaveBeenCalled();
  // });
});
