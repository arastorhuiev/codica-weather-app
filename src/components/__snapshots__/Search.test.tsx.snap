// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should create Search 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Enter city name
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1a1fmpi-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id=":r0:"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          Enter city name
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1a1fmpi-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id=":r0:"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
